File: apps/ios/HaruBook/Features/Bookshelf/BookshelfViewModel.swift
import Foundation
import Observation

@Observable
@MainActor
final class BookshelfViewModel {
    let profileId: Int
    var books: [Book] = []
    var cefrFilter: CefrLevel?
    var isLoading: Bool = false
    var error: String?
    var credits: CreditBalance?

    init(profileId: Int) {
        self.profileId = profileId
    }

    func fetchCredits() async {
        do {
            let response: CreditsResponse = try await APIClient.shared.send(
                Endpoint(path: "/api/billing/credits", method: .get)
            )
            self.credits = response.credits
        } catch {
            // 소프트 페일 — 별 잔액은 보조 정보.
        }
    }

    func load() async {
        isLoading = true
        defer { isLoading = false }

        var query: [URLQueryItem] = [
            URLQueryItem(name: "profileId", value: String(profileId))
        ]
        if let cefr = cefrFilter {
            query.append(URLQueryItem(name: "cefr", value: cefr.rawValue))
        }

        do {
            let response: BooksResponse = try await APIClient.shared.send(
                Endpoint(path: "/api/books", method: .get, query: query)
            )
            self.books = response.books
            self.error = nil
        } catch {
            self.error = error.localizedDescription
        }
    }

    func setCefr(_ cefr: CefrLevel?) async {
        cefrFilter = cefr
        await load()
    }
}

File: apps/android/app/src/main/java/site/smap/harubook/features/bookshelf/BookshelfViewModel.kt
package site.smap.harubook.features.bookshelf

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import site.smap.harubook.core.models.Book
import site.smap.harubook.core.models.BooksResponse
import site.smap.harubook.core.models.CefrLevel
import site.smap.harubook.core.models.CreditBalance
import site.smap.harubook.core.models.CreditsResponse
import site.smap.harubook.core.networking.ApiClient

data class BookshelfUiState(
    val books: List<Book> = emptyList(),
    val cefrFilter: CefrLevel? = null,
    val credits: CreditBalance? = null,
    val isLoading: Boolean = false,
    val error: String? = null,
)

class BookshelfViewModel(private val profileId: Int) : ViewModel() {
    private val _state = MutableStateFlow(BookshelfUiState())
    val state: StateFlow<BookshelfUiState> = _state.asStateFlow()

    fun load() {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true) }
            try {
                val query = buildMap {
                    put("profileId", profileId.toString())
                    _state.value.cefrFilter?.let { put("cefr", it.name) }
                }
                val response: BooksResponse = ApiClient.get(path = "/api/books", query = query)
                _state.update { it.copy(books = response.books, isLoading = false, error = null) }
            } catch (e: Throwable) {
                _state.update { it.copy(isLoading = false, error = e.message ?: "책장을 불러오지 못했어요.") }
            }
        }
    }

    fun fetchCredits() {
        viewModelScope.launch {
            try {
                val response: CreditsResponse = ApiClient.get("/api/billing/credits")
                _state.update { it.copy(credits = response.credits) }
            } catch (_: Throwable) {
                // 별 잔액은 보조 정보 — soft fail.
            }
        }
    }

    fun setCefr(cefr: CefrLevel?) {
        _state.update { it.copy(cefrFilter = cefr) }
        load()
    }
}
